Add randomization of XContentBuilder output to query tests

cbuescher opened the pull request:
Currently our testing of parsing query builders is limited to the default order of the parameters that each builders toXContent() method produces. To better test real queries where the order of parameters can be different, this change adds a helper method to ESTestCase that takes a XContentBuilder and randomly shuffles the order of the fields inside an object. This is now used in AbstractQueryTestCase, but it can be used in other similar places in the future.

cbuescher committed:
Add randomization of XContentBuilder output to query tests

Currently our testing of parsing query builders is limited to the
default order of the parameters that each builders toXContent()
method produces. To better test real queries where the order of
parameters can be different, this change adds a helper
method to ESTestCase that takes a XContentBuilder and randomly
shuffles the order of the fields inside an object. This is
used in AbstractQueryTestCase, but it can be used in other similar
places in the future.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -598,6 +605,30 @@ private static String groupName(ThreadGroup threadGroup) {
+
+    private static Map<String, Object> shuffleMap(Map<String, Object> map, Set<String> exceptFieldNames, Random r) {
+        List<String> keys = new ArrayList<>(map.keySet());
+        Map<String, Object> targetMap = new TreeMap<>();
```
I think you have to sort before shuffling, otherwise it will not be deterministic, cause the order in the list still depends on the order in the set, which depends on the jvm :)

nik9000 commented on core/src/test/java/org/elasticsearch/index/query/PercolatorQueryBuilderTests.java:
```diff
@@ -75,6 +78,18 @@ private PercolatorQueryBuilder doCreateTestQueryBuilder(boolean indexedDocument)
+    protected Set<String> provideShuffleproofFields() {
+        Set<String> fieldNames = new HashSet<>();
+        fieldNames.add(PercolatorQueryParser.DOCUMENT_FIELD.getPreferredName());
+        return fieldNames;
```
Collections.singleton?

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -598,6 +605,30 @@ private static String groupName(ThreadGroup threadGroup) {
+        XContentParser parser = XContentFactory.xContent(bytes).createParser(bytes);
+        // use ordered maps for reproducibility
+        Map<String, Object> shuffledMap = shuffleMap(parser.mapOrdered(), exceptFieldNames, random());
+        XContentBuilder jsonBuilder = XContentFactory.jsonBuilder();
```
Is it a good idea to always do json?

javanna replied:
it should rather save the XContentType detected from the bytes (XContentFactory.xContent(bytes)) above and re-use it.

nik9000 commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -598,6 +605,30 @@ private static String groupName(ThreadGroup threadGroup) {
+        return jsonBuilder.map(shuffledMap);
+    }
+
+    private static Map<String, Object> shuffleMap(Map<String, Object> map, Set<String> exceptFieldNames, Random r) {
```
Why pass random here?

javanna commented on core/src/test/java/org/elasticsearch/index/query/AbstractQueryTestCase.java:
```diff
@@ -390,6 +392,14 @@ public void testFromXContent() throws IOException {
+     * subclasses should override this method in case some fields in xContent should be protected from random
+     * shuffling in the {@link #testFromXContent()} test case
+     */
+    protected Set<String> provideShuffleproofFields() {
```
I expected we need to override this method for other cases. like I have memories of indices query depending on the order of keys. or did we fix that as part of the query refactoring?

cbuescher replied:
I also thought so, but at least for queries - with the exception of PercolatorQueryBuilder - random ordering of fields worked well. I ran the test suite several times, see what else CI brings up. But this way will detect those things.

javanna commented on core/src/test/java/org/elasticsearch/index/query/PercolatorQueryBuilderTests.java:
```diff
@@ -75,6 +78,18 @@ private PercolatorQueryBuilder doCreateTestQueryBuilder(boolean indexedDocument)
+     * BytesReference
+     */
+    @Override
+    protected Set<String> provideShuffleproofFields() {
```
the ones returned here are the fields that can be shuffled or that should not be shuffled? Not sure if the name is consistent with the behaviour

cbuescher committed:
Addressing review comments

cbuescher committed:
Improve some test method comments.

cbuescher commented:
@nik9000 @javanna thanks for the suggestions, I incorporated them into this PR.

martijnvg commented on core/src/test/java/org/elasticsearch/index/query/PercolatorQueryBuilderTests.java:
```diff
@@ -75,6 +79,16 @@ private PercolatorQueryBuilder doCreateTestQueryBuilder(boolean indexedDocument)
         }
     }
 
+    /**
```
I don't understand why comparing `BytesReference` would fail if the content is the same. Regardless of the location of `document` field, it is just parsing its content so it should always be the same? Or am I missing something here?

javanna replied:
the bytes are not the same right? I guess it depends on how we compare bytes references.... would be nice to get around it so we don't need this odd method at all :)

cbuescher replied:
Maybe the comment isn't clear hear. This PR changes the roundtrip test for xContent -> parsing -> builder so that all fields inside the query get randomly reordered. This also applies recursively to inner Objects. So when in the original builder you have ` "document" : { "foo" : "a", "bar" : "b" }` but you might get ` "document" : { "bar" : "b", "foo" : "a" }` in the parsed builder. Both documents are the same, but their bytes representation, which is stored in the builder, is different. This is why that fields inside "document" should not recursively be shuffled in the test.

martijnvg replied:
@cbuescher Thx, now I understand it is about the reordering of the fields inside the document itself. Maybe make sure that the comment emphasizes this?

cbuescher replied:
I improved the comment.

javanna commented on test/framework/src/main/java/org/elasticsearch/test/ESTestCase.java:
```diff
@@ -599,6 +605,38 @@ private static String groupName(ThreadGroup threadGroup) {
+    private static Map<String, Object> shuffleMap(Map<String, Object> map, Set<String> exceptFieldNames) {
+        List<String> keys = new ArrayList<>(map.keySet());
+        // even though we shuffle later, we need this to make tests reproduce on different jvms
+        //Collections.sort(keys);
```
this shouldn't be commented out :)

cbuescher replied:
d'oh

cbuescher commented:
@martijnvg @javanna improved the comment you mentioned. Can you take another look and let me know if theres anything left to add?

javanna commented:
LGTM @cbuescher . I am just not sure it should close #5831 as it shuffles things only for queries, while the original issue was more generic. Not sure it is doable though more generic than this and indeed it would be nice to close that old issue :)

cbuescher commented:
@javanna thanks, I removed closing #5831, but then we should revisit that issue and define more clearly what is required to close it. I'm planning to use this for the other parts of xContent in the search source, maybe we can close it when we added this kind of randomization to aggregations, sorts, suggesters, shapes etc. (to be defined in 5831)? Anyway, will merge this then.

martijnvg commented:
+1
